apiserver/facades/agent/uniter/uniter.go

```diff
@@ -2835,3 +2837,60 @@ func (u *UniterAPI) APIAddresses(ctx context.Context) (result params.StringsResu
 
 	return u.APIAddresser.APIAddresses(ctx, controllerConfig)
 }
+
+// Watch starts an NotifyWatcher for a unit or application.
+func (a *UniterAPI) Watch(ctx context.Context, args params.Entities) (params.NotifyWatchResults, error) {
+	result := params.NotifyWatchResults{
+		Results: make([]params.NotifyWatchResult, len(args.Entities)),
+	}
+	if len(args.Entities) == 0 {
+		return result, nil
+	}
+	canWatch, err := a.accessUnitOrApplication()
+	if err != nil {
+		return params.NotifyWatchResults{}, errors.Trace(err)
+	}
+	for i, entity := range args.Entities {
+		tag, err := names.ParseTag(entity.Tag)
+		if err != nil {
+			result.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
```

Comment:
**question:** Is it expected to return a ErrPerm and discard the underlying error ? It feels wrong, because it would make the analysis of this error path more difficult (ErrPerm is also used for an actual permission error)

**suggestion:** maybe decorate ErrPerm with the actual error, or join them.

Reply:
That makes sense, though in the past I've been told the ErrPerm is an exception to the rule of adding details to errors. If we're doing a permission denied error then we should give very little information away, since this could be an attack vector for juju. 

We don't annotate this error anywhere else in the uniter API, and this seems to be the standard way to do it in the uniter API. I agree with you in principle, but would rather do this as a wholesale change to the API rather than just here, i.e. if we were to add logging to it, or annotate it.